bson/bsoncodec/default_value_decoders.go

```diff
@@ -62,33 +62,66 @@ func (dvd DefaultValueDecoders) RegisterDefaultDecoders(rb *RegistryBuilder) {
 		RegisterDefaultDecoder(reflect.Struct, &StructCodec{cache: make(map[reflect.Type]*structDescription), parser: DefaultStructTagParser})
 }
 
+func notKindAndPtrToKind(val reflect.Value, kind reflect.Kind) bool {
+	if val.Type().Kind() == reflect.Ptr {
+		return val.Type().Elem().Kind() != kind
+	}
+	return val.Type().Kind() != kind
+}
+
 // BooleanDecodeValue is the ValueDecoderFunc for bool types.
 func (dvd DefaultValueDecoders) BooleanDecodeValue(dctx DecodeContext, vr bsonrw.ValueReader, i interface{}) error {
 	if vr.Type() != bsontype.Boolean {
 		return fmt.Errorf("cannot decode %v into a boolean", vr.Type())
 	}
 
-	var err error
-	if target, ok := i.(*bool); ok && target != nil { // if it is nil, we go the slow path.
-		*target, err = vr.ReadBoolean()
+	switch target := i.(type) {
+	case *bool:
+		if target == nil {
+			return errors.New("BooleanDecodeValue can only be used to decode settable (non-nil) values")
+		}
+		b, err := vr.ReadBoolean()
+		if err != nil {
+			return err
+		}
+		*target = b
 		return err
+	case **bool:
+		if target == nil {
+			return errors.New("BooleanDecodeValue can only be used to decode settable (non-nil) values")
+		}
+		b, err := vr.ReadBoolean()
+		if err != nil {
+			return err
+		}
+		*target = &b
 	}
 
 	val := reflect.ValueOf(i)
 	if !val.IsValid() || val.Kind() != reflect.Ptr || !val.Elem().CanSet() {
 		return errors.New("BooleanDecodeValue can only be used to decode settable (non-nil) values")
 	}
 	val = val.Elem()
-	if val.Type().Kind() != reflect.Bool {
+
+	if notKindAndPtrToKind(val, reflect.Bool) {
 		return ValueDecoderError{Name: "BooleanDecodeValue", Types: []interface{}{bool(true)}, Received: i}
 	}
 
 	b, err := vr.ReadBoolean()
-	val.SetBool(b)
+	if err != nil {
+		return err
+	}
+
+	if val.Kind() == reflect.Ptr {
+		val.Set(reflect.New(val.Type().Elem()))
```

Comment:
We should only allocate a new value here if the val is `nil`.

Reply:
a new function `getValue` is introduced to check and allocate new value.  
do you have any idea of a better name ?